File: backend/src/auth/jwt.rs
use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation, Algorithm};
use serde::{Deserialize, Serialize};

const JWT_SECRET: &[u8] = b"blocksmith_super_secret_key";

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Claims {
    pub sub: String,
    pub exp: usize,
    pub email: Option<String>,
}

pub async fn verify_jwt(token: &str) -> Result<Claims, String> {
    let validation = Validation::new(Algorithm::HS256);

    let token_data = decode::<Claims>(
        token,
        &DecodingKey::from_secret(JWT_SECRET),
        &validation,
    )
    .map_err(|e| e.to_string())?;

    Ok(token_data.claims)
}

pub fn create_jwt(subject: &str) -> Result<String, jsonwebtoken::errors::Error> {
    let claims = Claims {
        sub: subject.to_string(),
        exp: 2000000000,
        email: None,
    };
    encode(
        &Header::default(),
        &claims,
        &EncodingKey::from_secret(JWT_SECRET),
    )
}

File: backend/src/auth/jwt.ts
import { SignJWT, jwtVerify } from "jose";

export interface Claims {
  sub: string;
  exp: number;
}

export async function createJWT(secret: string, userId: string): Promise<string> {
  const key = new TextEncoder().encode(secret);

  return new SignJWT({ sub: userId })
    .setProtectedHeader({ alg: "HS256" })
    .setIssuedAt()
    .setExpirationTime("7d")
    .sign(key);
}

export async function verifyJWT(secret: string, token: string): Promise<Claims> {
  const key = new TextEncoder().encode(secret);

  const { payload } = await jwtVerify(token, key, {
    algorithms: ["HS256"],
  });

  return payload as unknown as Claims;
}
